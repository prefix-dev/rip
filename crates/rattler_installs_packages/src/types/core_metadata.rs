// Implementation comes from https://github.com/njsmith/posy/blob/main/src/vocab/core_metadata.rs
// Licensed under MIT or Apache-2.0

use super::extra::ParseExtraError;
use crate::{
    types::Extra, types::PackageName, types::ParsePackageNameError, types::RFC822ish,
    types::Version, types::VersionSpecifiers,
};
use once_cell::sync::Lazy;
use pep440_rs::Pep440Error;
use pep508_rs::Requirement;
use std::{collections::HashSet, str::FromStr};
use thiserror::Error;

#[derive(Debug, Clone)]

/// The core metadata of a wheel.
pub struct WheelCoreMetadata {
    /// The name of the package
    pub name: PackageName,
    /// Version w.r.t to PEP440
    pub version: Version,
    /// Version of the metadata
    pub metadata_version: MetadataVersion,
    /// Requirements for this distribution
    /// Matches the Requires-Dist field
    pub requires_dist: Vec<Requirement>,
    /// Python requirement
    pub requires_python: Option<VersionSpecifiers>,
    /// Extras provided by this distribution
    pub extras: HashSet<Extra>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
/// Wrapper around a PEP440 version
/// specifically for the metadata version
pub struct MetadataVersion(pub Version);

impl MetadataVersion {
    /// We consider that this implements PEP643
    /// if the version is 2.3 or higher.
    pub fn implements_pep643(&self) -> bool {
        static VERSION_2_2: Lazy<MetadataVersion> = Lazy::new(|| {
            MetadataVersion(Version::from_str("2.2").expect("cannot parse 2.2 version string"))
        });

        if self < &VERSION_2_2 {
            return false;
        }
        true
    }
}

#[derive(Debug, Error)]
#[allow(missing_docs)]
pub enum WheelCoreMetaDataError {
    #[error(transparent)]
    FailedToParseMetadata(#[from] <RFC822ish as FromStr>::Err),

    #[error("missing key {0} in METADATA")]
    MissingKey(String),

    #[error("duplicate key {0} in METADATA")]
    DuplicateKey(String),

    #[error("invalid Metadata-Version: {0}")]
    InvalidMetadataVersion(String),

    #[error("invalid Version: {0}")]
    InvalidVersion(String),

    #[error("invalid Requires-Python: {0}")]
    InvalidRequiresPython(#[source] Pep440Error),

    #[error("unsupported METADATA version {0}")]
    UnsupportedVersion(Version),

    #[error(transparent)]
    InvalidPackageName(#[from] ParsePackageNameError),

    #[error("invalid extra identifier '{0}'")]
    InvalidExtra(String, #[source] ParseExtraError),

    #[error("{0}")]
    FailedToParse(String),
}

impl TryFrom<&[u8]> for WheelCoreMetadata {
    type Error = WheelCoreMetaDataError;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        let (name, version, metadata_version, mut parsed) = parse_common(value)?;

        let mut requires_dist = Vec::new();
        for req_str in parsed.take_all("Requires-Dist").into_iter() {
            match req_str.parse() {
                Err(e) => {
                    tracing::warn!("ignoring Requires-Dist: {req_str}, failed to parse: {e}")
                }
                Ok(req) => requires_dist.push(req),
            }
        }

        let requires_python = parsed
            .maybe_take("Requires-Python")
            .map_err(|_| WheelCoreMetaDataError::DuplicateKey(String::from("Requires-Python")))?
            .as_deref()
            .map(VersionSpecifiers::from_str)
            .transpose()
            .map_err(WheelCoreMetaDataError::InvalidRequiresPython)?;

        let mut extras: HashSet<Extra> = HashSet::new();
        for extra in parsed.take_all("Provides-Extra").drain(..) {
            extras.insert(
                extra
                    .parse()
                    .map_err(|e| WheelCoreMetaDataError::InvalidExtra(extra, e))?,
            );
        }

        Ok(WheelCoreMetadata {
            name,
            version,
            metadata_version,
            requires_dist,
            requires_python,
            extras,
        })
    }
}

fn parse_common(
    input: &[u8],
) -> Result<(PackageName, Version, MetadataVersion, RFC822ish), WheelCoreMetaDataError> {
    let input = String::from_utf8_lossy(input);
    let mut parsed = RFC822ish::from_str(&input)?;

    static NEXT_MAJOR_METADATA_VERSION: Lazy<Version> =
        Lazy::new(|| Version::from_str("3").unwrap());

    // Quoth https://packaging.python.org/specifications/core-metadata:
    // "Automated tools consuming metadata SHOULD warn if metadata_version
    // is greater than the highest version they support, and MUST fail if
    // metadata_version has a greater major version than the highest
    // version they support (as described in PEP 440, the major version is
    // the value before the first dot)."
    //
    // We do the MUST, but I think I disagree about warning on
    // unrecognized minor revisions. If it's a minor revision, then by
    // definition old software is supposed to be able to handle it "well
    // enough". The only purpose of the warning would be to alert users
    // that they might want to upgrade, or to alert the tool authors that
    // there's a new metadata release. But for users, there are better
    // ways to nudge them to upgrade (e.g. checking on startup, like
    // pip does), and new metadata releases are so rare and so
    // much-discussed beforehand that if a tool's authors don't know
    // about it it's because the tool is abandoned anyway.
    let metadata_version = parsed
        .take("Metadata-Version")
        .map_err(|_| WheelCoreMetaDataError::MissingKey(String::from("Metadata-Version")))?;
    let metadata_version: Version = metadata_version
        .parse()
        .map_err(WheelCoreMetaDataError::InvalidMetadataVersion)?;
    if metadata_version >= *NEXT_MAJOR_METADATA_VERSION {
        return Err(WheelCoreMetaDataError::UnsupportedVersion(metadata_version));
    }

    let version_str = parsed
        .take("Version")
        .map_err(|_| WheelCoreMetaDataError::MissingKey(String::from("Version")))?;

    Ok((
        parsed
            .take("Name")
            .map_err(|_| WheelCoreMetaDataError::MissingKey(String::from("Name")))?
            .parse()?,
        version_str
            .parse()
            .map_err(WheelCoreMetaDataError::InvalidVersion)?,
        MetadataVersion(metadata_version),
        parsed,
    ))
}
