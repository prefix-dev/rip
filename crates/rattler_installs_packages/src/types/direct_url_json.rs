use serde::{Deserialize, Serialize};
use url::Url;

/// Specifies the PyPa `direct_url.json` format.
/// See: <https://packaging.python.org/en/latest/specifications/direct-url-data-structure/>
///
#[derive(Debug, Serialize, Deserialize)]
#[serde_with::skip_serializing_none]
pub struct DirectUrlJson {
    /// Url to the source.
    pub url: Url,
    /// Information about the source.
    #[serde(flatten)]
    pub source: DirectUrlSource,
}

/// Specifies the source of a direct url.
///
/// currently we do not support the deprecated `hash` field
#[derive(Debug, Serialize, Deserialize)]
pub enum DirectUrlSource {
    #[serde(rename = "archive_info")]
    /// Information about the archive file.
    Archive {
        /// Hashes of the archive file.
        hashes: DirectUrlHashes,
    },
    /// Information about a source from a VCS directly
    #[serde(rename = "vcs_info")]
    Vcs {
        /// The VCS used
        vcs: DirectUrlVcs,
        /// Revision of the source
        requested_revision: Option<String>,
        /// Actual commit
        commit_id: String,
    },
    /// Information about a local directory source
    #[serde(rename = "dir_info")]
    Dir {
        /// Is this a editable source
        /// See: <https://pip.pypa.io/en/stable/topics/local-project-installs/>
        editable: Option<bool>,
    },
}

/// Hashes for internal archive files.
/// multiple hashes can be included but per recommendation only sha256 should be used.
#[derive(Debug, Serialize, Deserialize)]
pub struct DirectUrlHashes {
    /// Sha256 hash of the archive file.
    pub sha256: String,
}

/// Name of the VCS in a DirectUrlSource
#[derive(Debug, Serialize, Deserialize)]
#[allow(missing_docs)]
pub enum DirectUrlVcs {
    #[serde(rename = "git")]
    Git,
    #[serde(rename = "svn")]
    Svn,
    #[serde(rename = "bzr")]
    Bazaar,
    #[serde(rename = "hg")]
    Mercurial,
}

#[cfg(test)]
mod tests {
    use crate::types::direct_url_json::DirectUrlJson;

    /// Tests if json outputs aligns with the examples at:
    /// https://packaging.python.org/en/latest/specifications/direct-url-data-structure/
    /// try to parse the example cases from there
    #[test]
    pub fn test_examples_pypa() {
        // Source archive:
        let example = r#"
        {
            "url": "https://github.com/pypa/pip/archive/1.3.1.zip",
            "archive_info": {
                "hashes": {
                    "sha256": "2dc6b5a470a1bde68946f263f1af1515a2574a150a30d6ce02c6ff742fcc0db8"
                }
            }
        }
        "#;
        serde_json::from_str::<DirectUrlJson>(example).unwrap();

        // Git URL with tag and commit-hash:
        let example = r#"
        {
            "url": "https://github.com/pypa/pip.git",
            "vcs_info": {
                "vcs": "git",
                "requested_revision": "1.3.1",
                "commit_id": "7921be1537eac1e97bc40179a57f0349c2aee67d"
            }
        }
        "#;
        serde_json::from_str::<DirectUrlJson>(example).unwrap();

        // Local directory:
        let example = r#"
        {
            "url": "file:///home/user/project",
            "dir_info": {}
        }
        "#;
        serde_json::from_str::<DirectUrlJson>(example).unwrap();

        // Local directory in editable mode:
        let example = r#"
        {
            "url": "file:///home/user/project",
            "dir_info": {
                "editable": true
            }
        }
        "#;
        serde_json::from_str::<DirectUrlJson>(example).unwrap();
    }
}
