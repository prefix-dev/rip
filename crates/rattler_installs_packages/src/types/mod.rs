//! This module contains all the types for working with PyPA packaging repositories.
//! We have tried to follow the PEP's and PyPA packaging guide as closely as possible.
mod artifact;

mod artifact_name;

mod package_name;

mod core_metadata;

mod record;

mod extra;

mod entry_points;

mod project_info;

mod direct_url_json;
mod rfc822ish;

pub use artifact::Artifact;

pub use artifact_name::{
    ArtifactName, BuildTag, InnerAsArtifactName, ParseArtifactNameError, SDistFilename,
    SDistFormat, WheelFilename,
};

pub use direct_url_json::{DirectUrlHashes, DirectUrlJson, DirectUrlSource, DirectUrlVcs};

pub use core_metadata::{MetadataVersion, WheelCoreMetaDataError, WheelCoreMetadata};

pub use record::{Record, RecordEntry};

pub use package_name::{NormalizedPackageName, PackageName, ParsePackageNameError};

pub use extra::Extra;

pub use entry_points::{EntryPoint, ParseEntryPointError};

pub use project_info::{ArtifactHashes, ArtifactInfo, DistInfoMetadata, Meta, ProjectInfo, Yanked};

pub(crate) use rfc822ish::RFC822ish;

pub use pep440_rs::{Version, VersionSpecifier, VersionSpecifiers};
pub use pep508_rs::{MarkerEnvironment, Requirement};
