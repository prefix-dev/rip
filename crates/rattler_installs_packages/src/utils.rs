use futures::{AsyncRead, AsyncReadExt, AsyncSeekExt};
use include_dir::{include_dir, Dir};
use std::io::{Read, Seek, SeekFrom};
use tokio_util::compat::TokioAsyncReadCompatExt;
use url::Url;

/// Keep retrying a certain IO function until it either succeeds or until it doesn't return
/// [`std::io::ErrorKind::Interrupted`].
pub fn retry_interrupted<F, T>(mut f: F) -> std::io::Result<T>
where
    F: FnMut() -> std::io::Result<T>,
{
    loop {
        match f() {
            Ok(result) => return Ok(result),
            Err(err) if err.kind() != std::io::ErrorKind::Interrupted => {
                return Err(err);
            }
            _ => {
                // Otherwise keep looping!
            }
        }
    }
}

/// Represents either data coming from the network in an async fashion or a local thing on disk.
/// We only use async for the network stuff, the local filesystem doesn't really benefit from it.
pub enum StreamingOrLocal {
    Streaming(Box<dyn AsyncRead + Unpin + Send>),
    Local(Box<dyn ReadAndSeek + Send>),
}

pub trait ReadAndSeek: Read + Seek {}
impl<T> ReadAndSeek for T where T: Read + Seek {}

impl StreamingOrLocal {
    /// Returns an instance that is both readable and seekable by first streaming the contents to
    /// disk if required.
    pub async fn force_local(self) -> std::io::Result<Box<dyn ReadAndSeek + Send>> {
        Ok(match self {
            StreamingOrLocal::Local(stream) => stream,
            StreamingOrLocal::Streaming(mut stream) => {
                let mut tmp = tokio::fs::File::from(tempfile::tempfile()?).compat();
                futures::io::copy(&mut stream, &mut tmp).await?;
                tmp.seek(SeekFrom::Start(0)).await?;
                Box::new(tmp.into_inner().into_std().await)
            }
        })
    }

    pub async fn read_to_end(self, bytes: &mut Vec<u8>) -> std::io::Result<usize> {
        match self {
            StreamingOrLocal::Streaming(mut streaming) => streaming.read_to_end(bytes).await,
            StreamingOrLocal::Local(mut local) => {
                match tokio::task::spawn_blocking(move || {
                    let mut bytes = Vec::new();
                    local.read_to_end(&mut bytes).map(|_| bytes)
                })
                .await
                {
                    Ok(Ok(result)) => {
                        *bytes = result;
                        Ok(bytes.len())
                    }
                    Ok(Err(err)) => Err(err),
                    Err(err) => {
                        if let Ok(panic) = err.try_into_panic() {
                            std::panic::resume_unwind(panic)
                        }
                        Err(std::io::ErrorKind::Interrupted.into())
                    }
                }
            }
        }
    }
}

/// Normalize url according to pip standards
pub fn normalize_index_url(mut url: Url) -> Url {
    let path = url.path();
    if !path.ends_with('/') {
        url.set_path(&format!("{path}/"));
    }
    url
}

pub(crate) static VENDORED_PACKAGING_DIR: Dir<'_> =
    include_dir!("$CARGO_MANIFEST_DIR/vendor/packaging/");
