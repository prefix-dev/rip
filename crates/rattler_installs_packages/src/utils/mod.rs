mod read_and_seek;
mod streaming_or_local;

mod seek_slice;

use include_dir::{include_dir, Dir};
use url::Url;

pub use read_and_seek::ReadAndSeek;
pub use streaming_or_local::StreamingOrLocal;

pub use seek_slice::SeekSlice;

/// Keep retrying a certain IO function until it either succeeds or until it doesn't return
/// [`std::io::ErrorKind::Interrupted`].
pub fn retry_interrupted<F, T>(mut f: F) -> std::io::Result<T>
where
    F: FnMut() -> std::io::Result<T>,
{
    loop {
        match f() {
            Ok(result) => return Ok(result),
            Err(err) if err.kind() != std::io::ErrorKind::Interrupted => {
                return Err(err);
            }
            _ => {
                // Otherwise keep looping!
            }
        }
    }
}

/// Normalize url according to pip standards
pub fn normalize_index_url(mut url: Url) -> Url {
    let path = url.path();
    if !path.ends_with('/') {
        url.set_path(&format!("{path}/"));
    }
    url
}

pub(crate) static VENDORED_PACKAGING_DIR: Dir<'_> =
    include_dir!("$CARGO_MANIFEST_DIR/vendor/packaging/");
