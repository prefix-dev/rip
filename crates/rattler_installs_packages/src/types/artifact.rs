use super::artifact_name::InnerAsArtifactName;
use crate::resolve::PypiVersion;
use crate::types::SourceArtifactName;
use crate::utils::ReadAndSeek;
use std::path::Path;

/// Trait to implement if it is a type that has an [`super::artifact_name::ArtifactName`]
/// this is then used by the [`crate::index::PackageDb`] to make a difference
/// between the different types of artifacts.
pub trait HasArtifactName {
    /// The name of the artifact which describes the artifact.
    ///
    /// Artifacts are describes by a string. [`super::artifact_name::ArtifactName`] describes the
    /// general format.
    type Name: Clone + InnerAsArtifactName;

    /// Returns the name of this instance
    fn name(&self) -> &Self::Name;
}

/// Trait that represents an artifact type in the PyPI ecosystem.
/// That is a single file like a wheel, sdist.
pub trait ArtifactFromBytes: HasArtifactName + Sized {
    /// Construct a new artifact from the given bytes
    fn from_bytes(name: Self::Name, bytes: Box<dyn ReadAndSeek + Send>) -> miette::Result<Self>;
}

/// Error while reading pyproject.toml
#[derive(thiserror::Error, Debug)]
#[allow(missing_docs)]
pub enum ReadPyProjectError {
    #[error("IO error while reading pyproject.toml: {0}")]
    Io(#[from] std::io::Error),

    #[error("No pyproject.toml found in archive")]
    NoPyProjectTomlFound,

    #[error("Could not parse pyproject.toml")]
    PyProjectTomlParseError(String),
}

/// SDist or STree act as a SourceArtifact
/// so we can use it in methods where we expect sdist
/// to extract metadata
pub trait ArtifactFromSource: HasArtifactName + Sync {
    /// get bytes of an artifact
    /// that will we be used for hashing
    fn try_get_bytes(&self) -> Result<Vec<u8>, std::io::Error>;

    /// Distribution Name
    fn distribution_name(&self) -> String;

    /// Version ( URL or Version )
    fn version(&self) -> PypiVersion;

    /// Source artifact name
    fn artifact_name(&self) -> SourceArtifactName;

    /// Read the build system info from the pyproject.toml
    fn read_build_info(&self) -> Result<pyproject_toml::BuildSystem, ReadPyProjectError>;

    /// extract to a specific location
    /// for sdist we unpack it
    /// for stree we move it
    /// as example this method is used by install_build_files
    fn extract_to(&self, work_dir: &Path) -> std::io::Result<()>;
}
