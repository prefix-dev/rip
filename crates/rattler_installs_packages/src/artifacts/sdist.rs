use crate::types::{Artifact, NormalizedPackageName, SDistFilename, SDistFormat};
use crate::types::{WheelCoreMetaDataError, WheelCoreMetadata};
use crate::utils::ReadAndSeek;
use flate2::read::GzDecoder;
use miette::IntoDiagnostic;
use parking_lot::{Mutex, MutexGuard};
use serde::Serialize;

use fs_err as fs;
use std::ffi::OsStr;
use std::io::{ErrorKind, Read, Seek};
use std::path::{Path, PathBuf};
use tar::Archive;

/// Represents a source distribution artifact.
pub struct SDist {
    /// Name of the source distribution
    name: SDistFilename,

    /// Source dist archive
    file: Mutex<Box<dyn ReadAndSeek + Send>>,
}

#[derive(Debug, Serialize)]
pub struct BuildSystem {
    backend_path: Vec<PathBuf>,
    build_backend: String,
}

#[derive(thiserror::Error, Debug)]
pub enum SDistError {
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    #[error("No PKG-INFO found in archive")]
    NoPkgInfoFound,

    #[error("No pyproject.toml found in archive")]
    NoPyProjectTomlFound,

    #[error("Could not parse pyproject.toml")]
    PyProjectTomlParseError(String),

    #[error("Could not parse metadata")]
    WheelCoreMetaDataError(#[from] WheelCoreMetaDataError),
}

impl SDist {
    /// Create this struct from a path
    #[allow(dead_code)]
    pub fn from_path(
        path: &Path,
        normalized_package_name: &NormalizedPackageName,
    ) -> miette::Result<Self> {
        let file_name = path
            .file_name()
            .and_then(OsStr::to_str)
            .ok_or_else(|| miette::miette!("path does not contain a filename"))?;
        let name =
            SDistFilename::from_filename(file_name, normalized_package_name).into_diagnostic()?;
        let bytes = fs::File::open(path).into_diagnostic()?;
        Self::new(name, Box::new(bytes))
    }

    /// Find entry in tar archive
    fn find_entry(&self, name: impl AsRef<Path>) -> std::io::Result<Option<Vec<u8>>> {
        let mut lock = self.file.lock();
        let mut archive = generic_archive_reader(&mut lock, self.name.format)?;

        fn skip_first_component(path: &Path) -> PathBuf {
            path.components().skip(1).collect()
        }

        // Loop over entries
        for entry in archive.entries()? {
            let mut entry = entry?;

            // Find name in archive and return this
            if skip_first_component(entry.path()?.as_ref()) == name.as_ref() {
                let mut bytes = Vec::new();
                entry.read_to_end(&mut bytes)?;
                return Ok(Some(bytes));
            }
        }
        Ok(None)
    }

    /// Read .PKG-INFO from the archive
    pub fn read_package_info(&self) -> Result<(Vec<u8>, WheelCoreMetadata), SDistError> {
        if let Some(bytes) = self.find_entry("PKG-INFO")? {
            let metadata = WheelCoreMetadata::try_from(bytes.as_slice())?;

            Ok((bytes, metadata))
        } else {
            Err(SDistError::NoPkgInfoFound)
        }
    }

    /// Read the build system info from the pyproject.toml
    pub fn read_build_info(&self) -> Result<pyproject_toml::BuildSystem, SDistError> {
        if let Some(bytes) = self.find_entry("pyproject.toml")? {
            let source = String::from_utf8(bytes).map_err(|e| {
                SDistError::PyProjectTomlParseError(format!(
                    "could not parse pyproject.toml (bad encoding): {}",
                    e
                ))
            })?;
            let project = pyproject_toml::PyProjectToml::new(&source).map_err(|e| {
                SDistError::PyProjectTomlParseError(format!(
                    "could not parse pyproject.toml (bad toml): {}",
                    e
                ))
            })?;
            Ok(project
                .build_system
                .ok_or_else(|| std::io::Error::new(ErrorKind::NotFound, "no build-system found"))?)
        } else {
            Err(SDistError::NoPyProjectTomlFound)
        }
    }

    /// Extract the contents of the sdist archive to the given directory
    pub fn extract_to(&self, work_dir: &Path) -> std::io::Result<()> {
        let mut lock = self.file.lock();
        let mut archive = generic_archive_reader(&mut lock, self.name.format)?;
        archive.unpack(work_dir)?;
        Ok(())
    }

    /// Checks if this artifact implements PEP 643
    /// and returns the metadata if it does
    pub fn pep643_metadata(&self) -> Result<Option<(Vec<u8>, WheelCoreMetadata)>, SDistError> {
        // Assume we have a PKG-INFO
        let (bytes, metadata) = self.read_package_info()?;
        if metadata.metadata_version.implements_pep643() {
            Ok(Some((bytes, metadata)))
        } else {
            Ok(None)
        }
    }

    /// Get a lock on the inner data
    pub fn lock_data(&self) -> MutexGuard<Box<dyn ReadAndSeek + Send>> {
        self.file.lock()
    }
}

impl Artifact for SDist {
    type Name = SDistFilename;

    fn new(name: Self::Name, bytes: Box<dyn ReadAndSeek + Send>) -> miette::Result<Self> {
        Ok(Self {
            name,
            file: Mutex::new(bytes),
        })
    }

    fn name(&self) -> &Self::Name {
        &self.name
    }
}

enum RawAndGzReader<'a> {
    Raw(&'a mut Box<dyn ReadAndSeek + Send>),
    Gz(GzDecoder<&'a mut Box<dyn ReadAndSeek + Send>>),
}

impl<'a> Read for RawAndGzReader<'a> {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        match self {
            Self::Raw(r) => r.read(buf),
            Self::Gz(r) => r.read(buf),
        }
    }
}

fn generic_archive_reader(
    file: &mut Box<dyn ReadAndSeek + Send>,
    format: SDistFormat,
) -> std::io::Result<Archive<RawAndGzReader>> {
    file.rewind()?;

    match format {
        SDistFormat::TarGz => {
            let bytes = GzDecoder::new(file);
            Ok(Archive::new(RawAndGzReader::Gz(bytes)))
        }
        SDistFormat::Tar => Ok(Archive::new(RawAndGzReader::Raw(file))),
        _ => Err(std::io::Error::new(
            ErrorKind::InvalidData,
            "sdist archive format currently unsupported (only tar and tar.gz are supported)",
        )),
    }
}

#[cfg(test)]
mod tests {
    use crate::artifacts::SDist;
    use crate::python_env::Pep508EnvMakers;
    use crate::wheel_builder::WheelBuilder;
    use crate::{index::PackageDb, resolve::ResolveOptions};
    use insta::{assert_debug_snapshot, assert_ron_snapshot};
    use std::collections::HashMap;
    use std::path::Path;
    use tempfile::TempDir;

    fn get_package_db() -> (PackageDb, TempDir) {
        let tempdir = tempfile::tempdir().unwrap();
        (
            PackageDb::new(
                Default::default(),
                &[url::Url::parse("https://pypi.org/simple/").unwrap()],
                tempdir.path(),
            )
            .unwrap(),
            tempdir,
        )
    }

    #[tokio::test]
    pub async fn correct_metadata_fake_flask() {
        let path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("../../test-data/sdists/fake-flask-3.0.0.tar.gz");

        let sdist = SDist::from_path(&path, &"fake-flask".parse().unwrap()).unwrap();
        // Should not fail as it is a valid PKG-INFO
        // and considered reliable
        let _package_db = get_package_db();
        sdist.pep643_metadata().unwrap().unwrap();
    }

    #[test]
    pub fn read_rich_build_info() {
        // Read path
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/sdists/rich-13.6.0.tar.gz");

        // Load sdist
        let sdist = super::SDist::from_path(&path, &"rich".parse().unwrap()).unwrap();

        let build_system = sdist.read_build_info().unwrap();

        assert_ron_snapshot!(build_system, @r###"
        BuildSystem(
          requires: [
            "poetry-core >=1.0.0",
          ],
          r#build-backend: Some("poetry.core.masonry.api"),
          r#backend-path: None,
        )
        "###);
    }

    #[tokio::test(flavor = "multi_thread")]
    pub async fn sdist_metadata() {
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/sdists/rich-13.6.0.tar.gz");

        let sdist = SDist::from_path(&path, &"rich".parse().unwrap()).unwrap();

        let package_db = get_package_db();
        let env_markers = Pep508EnvMakers::from_env().await.unwrap();
        let resolve_options = ResolveOptions::default();
        let wheel_builder = WheelBuilder::new(
            &package_db.0,
            &env_markers,
            None,
            &resolve_options,
            package_db.1.path(),
            HashMap::default(),
        );

        let result = wheel_builder.get_sdist_metadata(&sdist).await.unwrap();

        assert_debug_snapshot!(result.1);
    }

    #[tokio::test(flavor = "multi_thread")]
    pub async fn build_rich_with_metadata() {
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/sdists/rich-13.6.0.tar.gz");

        let sdist = SDist::from_path(&path, &"rich".parse().unwrap()).unwrap();

        let package_db = get_package_db();
        let env_markers = Pep508EnvMakers::from_env().await.unwrap();
        let resolve_options = ResolveOptions::default();
        let wheel_builder = WheelBuilder::new(
            &package_db.0,
            &env_markers,
            None,
            &resolve_options,
            package_db.1.path(),
            HashMap::default(),
        );

        // Build the wheel
        wheel_builder.get_sdist_metadata(&sdist).await.unwrap();
        let wheel = wheel_builder.build_wheel(&sdist).await.unwrap();

        let (_, metadata) = wheel.metadata().unwrap();
        assert_debug_snapshot!(metadata);
    }
    #[tokio::test(flavor = "multi_thread")]
    pub async fn build_rich_no_metadata() {
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/sdists/rich-13.6.0.tar.gz");

        let sdist = SDist::from_path(&path, &"rich".parse().unwrap()).unwrap();

        let package_db = get_package_db();
        let env_markers = Pep508EnvMakers::from_env().await.unwrap();
        let resolve_options = ResolveOptions::default();
        let wheel_builder = WheelBuilder::new(
            &package_db.0,
            &env_markers,
            None,
            &resolve_options,
            package_db.1.path(),
            HashMap::default(),
        );

        // Build the wheel
        let wheel = wheel_builder.build_wheel(&sdist).await.unwrap();

        let (_, metadata) = wheel.metadata().unwrap();
        assert_debug_snapshot!(metadata);
    }

    #[tokio::test(flavor = "multi_thread")]
    pub async fn build_wheel_and_pass_env_variables() {
        let path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("../../test-data/sdists/env_package-0.1.tar.gz");

        let sdist = SDist::from_path(&path, &"env_package".parse().unwrap()).unwrap();

        let package_db = get_package_db();
        let env_markers = Pep508EnvMakers::from_env().await.unwrap();
        let resolve_options = ResolveOptions {
            ..Default::default()
        };

        let mut mandatory_env = HashMap::new();

        // In order to build wheel, we need to pass specific ENV that setup.py expect
        mandatory_env.insert("MY_ENV_VAR".to_string(), "SOME_VALUE".to_string());

        let wheel_builder = WheelBuilder::new(
            &package_db.0,
            &env_markers,
            None,
            &resolve_options,
            package_db.1.path(),
            mandatory_env,
        );

        // Build the wheel
        let wheel = wheel_builder.build_wheel(&sdist).await.unwrap();

        let (_, metadata) = wheel.metadata().unwrap();
        assert_debug_snapshot!(metadata);
    }

    // On windows these tests will fail because python interpreter
    // should have SYSTEMROOT
    // https://github.com/pyinstaller/pyinstaller/issues/6878
    #[cfg(not(target_os = "windows"))]
    #[tokio::test(flavor = "multi_thread")]
    pub async fn build_wheel_and_with_clean_env_and_pass_env_variables() {
        let path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("../../test-data/sdists/env_package-0.1.tar.gz");

        let sdist = SDist::from_path(&path, &"env_package".parse().unwrap()).unwrap();

        let package_db = get_package_db();
        let env_markers = Pep508EnvMakers::from_env().await.unwrap();
        let resolve_options = ResolveOptions {
            clean_env: true,
            ..Default::default()
        };

        let mut mandatory_env = HashMap::new();

        // In order to build wheel, we need to pass specific ENV that setup.py expect
        mandatory_env.insert(String::from("MY_ENV_VAR"), String::from("SOME_VALUE"));

        let wheel_builder = WheelBuilder::new(
            &package_db.0,
            &env_markers,
            None,
            &resolve_options,
            package_db.1.path(),
            mandatory_env,
        );

        // Build the wheel
        let wheel = wheel_builder.build_wheel(&sdist).await.unwrap();

        let (_, metadata) = wheel.metadata().unwrap();
        assert_debug_snapshot!(metadata);
    }

    #[cfg(not(target_os = "windows"))]
    #[tokio::test(flavor = "multi_thread")]
    pub async fn build_wheel_and_will_fail_when_clean_env_is_used() {
        let path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("../../test-data/sdists/env_package-0.1.tar.gz");

        let sdist = SDist::from_path(&path, &"env_package".parse().unwrap()).unwrap();

        let package_db = get_package_db();
        let env_markers = Pep508EnvMakers::from_env().await.unwrap();
        let resolve_options = ResolveOptions {
            clean_env: true,
            ..Default::default()
        };

        // Do not pass any mandatory env for wheel builder, and do not inherit
        // this should fail
        let mandatory_env = HashMap::new();

        let wheel_builder = WheelBuilder::new(
            &package_db.0,
            &env_markers,
            None,
            &resolve_options,
            package_db.1.path(),
            mandatory_env,
        );

        // Build the wheel
        let wheel = wheel_builder.build_wheel(&sdist).await;
        let err_string = wheel.err().unwrap().to_string();

        assert!(err_string.contains("could not build wheel"));
        assert!(err_string.contains("MY_ENV_VAR should be set in order to build wheel"));
    }
}
