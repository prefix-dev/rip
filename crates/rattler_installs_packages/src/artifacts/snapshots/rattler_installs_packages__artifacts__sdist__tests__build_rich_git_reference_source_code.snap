---
source: crates/rattler_installs_packages/src/artifacts/sdist.rs
expression: "artifact_info[0].filename"
---
STree(
    STreeFilename {
        distribution: PackageName {
            source: "rich",
            normalized: "rich",
        },
        version: Version {
            epoch: 0,
            release: [
                13,
                7,
                0,
            ],
            pre: None,
            post: None,
            dev: None,
            local: None,
        },
        url: Url {
            scheme: "git+https",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: Some(
                Domain(
                    "github.com",
                ),
            ),
            port: None,
            path: "/Textualize/rich.git@v13.7.0",
            query: None,
            fragment: None,
        },
    },
)
